File: paint/sources/material_nodes/mapping_node.ts

function mapping_node_init() {
	array_push(nodes_material_vector, mapping_node_def);
	map_set(parser_material_node_vectors, "MAPPING", mapping_node_vector);
}

function mapping_node_vector(node: ui_node_t, socket: ui_node_socket_t): string {
	let out: string              = parser_material_parse_vector_input(node.inputs[0]);
	let node_translation: string = parser_material_parse_vector_input(node.inputs[1]);
	let node_rotation: string    = parser_material_parse_vector_input(node.inputs[2]);
	let node_scale: string       = parser_material_parse_vector_input(node.inputs[3]);
	if (node_scale != "float3(1, 1, 1)") {
		out = "(" + out + " * " + node_scale + ")";
	}
	if (node_rotation != "float3(0, 0, 0)") {
		// ZYX rotation, Z axis for now..
		let a: string = node_rotation + ".z * (3.1415926535 / 180)";
		// x * cos(theta) - y * sin(theta)
		// x * sin(theta) + y * cos(theta)
		out = "float3(" + out + ".x * cos(" + a + ") - " + out + ".y * sin(" + a + "), " + out + ".x * sin(" + a + ") + " + out + ".y * cos(" + a + "), 0.0)";
	}
	// if node.rotation[1] != 0.0:
	//     a = node.rotation[1]
	//     out = "float3({0}.x * {1} - {0}.z * {2}, {0}.x * {2} + {0}.z * {1}, 0.0)".format(out, math_cos(a), math_sin(a))
	// if node.rotation[0] != 0.0:
	//     a = node.rotation[0]
	//     out = "float3({0}.y * {1} - {0}.z * {2}, {0}.y * {2} + {0}.z * {1}, 0.0)".format(out, math_cos(a), math_sin(a))
	if (node_translation != "float3(0, 0, 0)") {
		out = "(" + out + " + " + node_translation + ")";
	}
	// if node.use_min:
	// out = "max({0}, float3({1}, {2}, {3}))".format(out, node.min[0], node.min[1])
	// if node.use_max:
	// out = "min({0}, float3({1}, {2}, {3}))".format(out, node.max[0], node.max[1])
	return out;
}

let mapping_node_def: ui_node_t = {
	id : 0,
	name : _tr("Mapping"),
	type : "MAPPING",
	x : 0,
	y : 0,
	color : 0xff522c99,
	inputs : [
		{
			id : 0,
			node_id : 0,
			name : _tr("Vector"),
			type : "VECTOR",
			color : 0xff6363c7,
			default_value : f32_array_create_xyz(0.0, 0.0, 0.0),
			min : 0.0,
			max : 1.0,
			precision : 100,
			display : 0
		},
		{
			id : 0,
			node_id : 0,
			name : _tr("Location"),
			type : "VECTOR",
			color : 0xff6363c7,
			default_value : f32_array_create_xyz(0.0, 0.0, 0.0),
			min : 0.0,
			max : 1.0,
			precision : 100,
			display : 1
		},
		{
			id : 0,
			node_id : 0,
			name : _tr("Rotation"),
			type : "VECTOR",
			color : 0xff6363c7,
			default_value : f32_array_create_xyz(0.0, 0.0, 0.0),
			min : 0.0,
			max : 360.0,
			precision : 100,
			display : 1
		},
		{
			id : 0,
			node_id : 0,
			name : _tr("Scale"),
			type : "VECTOR",
			color : 0xff6363c7,
			default_value : f32_array_create_xyz(1.0, 1.0, 1.0),
			min : 0.0,
			max : 1.0,
			precision : 100,
			display : 1
		}
	],
	outputs : [ {
		id : 0,
		node_id : 0,
		name : _tr("Vector"),
		type : "VECTOR",
		color : 0xff6363c7,
		default_value : f32_array_create_xyz(0.0, 0.0, 0.0),
		min : 0.0,
		max : 1.0,
		precision : 100,
		display : 0
	} ],
	buttons : [],
	width : 0,
	flags : 0
};

File: paint/sources/material_nodes/mapping_node.c

#include "../global.h"

char *mapping_node_vector(ui_node_t *node, ui_node_socket_t *socket) {
	char *out              = parser_material_parse_vector_input(node->inputs->buffer[0]);
	char *node_translation = parser_material_parse_vector_input(node->inputs->buffer[1]);
	char *node_rotation    = parser_material_parse_vector_input(node->inputs->buffer[2]);
	char *node_scale       = parser_material_parse_vector_input(node->inputs->buffer[3]);
	if (!string_equals(node_scale, "float3(1.0, 1.0, 1.0)")) {
		out = string("(%s * %s)", out, node_scale);
	}
	if (!string_equals(node_rotation, "float3(0.0, 0.0, 0.0)")) {
		char *name = parser_material_store_var_name(node);
		char *v    = string("%s_v", name);
		char *rx   = string("%s_rx", name);
		char *ry   = string("%s_ry", name);
		char *rz   = string("%s_rz", name);
		parser_material_write(parser_material_kong, string("var %s: float3 = %s;", v, out));
		parser_material_write(parser_material_kong, string("var %s: float = %s.x * (3.14159265 / 180.0);", rx, node_rotation));
		parser_material_write(parser_material_kong, string("var %s: float = %s.y * (3.14159265 / 180.0);", ry, node_rotation));
		parser_material_write(parser_material_kong, string("var %s: float = %s.z * (3.14159265 / 180.0);", rz, node_rotation));
		parser_material_write(parser_material_kong,
		                      string("%s = float3(%s.x * cos(%s) - %s.y * sin(%s), %s.x * sin(%s) + %s.y * cos(%s), %s.z);", v, v, rz, v, rz, v, rz, v, rz, v));
		parser_material_write(parser_material_kong, string("%s = float3(%s.x * cos(%s) + %s.z * sin(%s), %s.y, -%s.x * sin(%s) + %s.z * cos(%s));", v, v, ry, v,
		                                                   ry, v, v, ry, v, ry));
		parser_material_write(parser_material_kong,
		                      string("%s = float3(%s.x, %s.y * cos(%s) - %s.z * sin(%s), %s.y * sin(%s) + %s.z * cos(%s));", v, v, v, rx, v, rx, v, rx, v, rx));
		out = v;
	}
	if (!string_equals(node_translation, "float3(0.0, 0.0, 0.0)")) {
		out = string("(%s + %s)", out, node_translation);
	}
	return out;
}

void mapping_node_init() {

	ui_node_t *mapping_node_def =
	    GC_ALLOC_INIT(ui_node_t, {.id     = 0,
	                              .name   = _tr("Mapping"),
	                              .type   = "MAPPING",
	                              .x      = 0,
	                              .y      = 0,
	                              .color  = 0xff522c99,
	                              .inputs = any_array_create_from_raw(
	                                  (void *[]){
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Vector"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(0.0, 0.0, 0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 1}),
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Location"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(0.0, 0.0, 0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 1}),
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Rotation"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(0.0, 0.0, 0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 360.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 1}),
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Scale"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(1.0, 1.0, 1.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 1}),
	                                  },
	                                  4),
	                              .outputs = any_array_create_from_raw(
	                                  (void *[]){
	                                      GC_ALLOC_INIT(ui_node_socket_t, {.id            = 0,
	                                                                       .node_id       = 0,
	                                                                       .name          = _tr("Vector"),
	                                                                       .type          = "VECTOR",
	                                                                       .color         = 0xff6363c7,
	                                                                       .default_value = f32_array_create_xyz(0.0, 0.0, 0.0),
	                                                                       .min           = 0.0,
	                                                                       .max           = 1.0,
	                                                                       .precision     = 100,
	                                                                       .display       = 0}),
	                                  },
	                                  1),
	                              .buttons = any_array_create_from_raw((void *[]){}, 0),
	                              .width   = 0,
	                              .flags   = 0});

	any_array_push(nodes_material_utilities, mapping_node_def);
	any_map_set(parser_material_node_vectors, "MAPPING", mapping_node_vector);
}
